File: javascript/test/Behavioral/Visitor.js
(function() {
  define(['Behavioral/Visitor/DvdInfo', 'Behavioral/Visitor/BookInfo', 'Behavioral/Visitor/GameInfo', 'Behavioral/Visitor/TitleLongBlurbVisitor', 'Behavioral/Visitor/TitleShortBlurbVisitor'], function(DvdInfo, BookInfo, GameInfo, TitleLongBlurbVisitor, TitleShortBlurbVisitor) {
    'use strict';
    return describe('Test Behavioral Visitor', function() {
      var bladeRunner, electricSheep, sheepRaider;
      bladeRunner = new DvdInfo("Blade Runner", "Harrison Ford", '1');
      electricSheep = new BookInfo("Do Androids Dream of Electric Sheep?", "Phillip K. Dick");
      sheepRaider = new GameInfo("Sheep Raider");
      it("Long Blurbs:", function() {
        var titleLongBlurbVisitor;
        titleLongBlurbVisitor = new TitleLongBlurbVisitor();
        bladeRunner.accept(titleLongBlurbVisitor);
        expect("Testing bladeRunner   " + titleLongBlurbVisitor.getTitleBlurb()).toEqual('Testing bladeRunner   LB-DVD: Blade Runner, starring Harrison Ford, encoding region: 1');
        electricSheep.accept(titleLongBlurbVisitor);
        expect("Testing electricSheep " + titleLongBlurbVisitor.getTitleBlurb()).toEqual('Testing electricSheep LB-Book: Do Androids Dream of Electric Sheep?, Author: Phillip K. Dick');
        sheepRaider.accept(titleLongBlurbVisitor);
        return expect("Testing sheepRaider   " + titleLongBlurbVisitor.getTitleBlurb()).toEqual('Testing sheepRaider   LB-Game: Sheep Raider');
      });
      return it("Short Blurbs:", function() {
        var titleShortBlurbVisitor;
        titleShortBlurbVisitor = new TitleShortBlurbVisitor();
        bladeRunner.accept(titleShortBlurbVisitor);
        expect("Testing bladeRunner   " + titleShortBlurbVisitor.getTitleBlurb()).toEqual("Testing bladeRunner   SB-DVD: Blade Runner");
        electricSheep.accept(titleShortBlurbVisitor);
        expect("Testing electricSheep " + titleShortBlurbVisitor.getTitleBlurb()).toEqual("Testing electricSheep SB-Book: Do Androids Dream of Electric Sheep?");
        sheepRaider.accept(titleShortBlurbVisitor);
        return expect("Testing sheepRaider   " + titleShortBlurbVisitor.getTitleBlurb()).toEqual('Testing sheepRaider   SB-Game: Sheep Raider');
      });
    });
  });

}).call(this);

File: clojure/src/patterns/behavioral/visitor.clj
(ns patterns.behavioral.visitor
   (:require [clojure.string :as str])
)

(defrecord Shortblurb [])
(defrecord Longblurb [])
(defrecord Bookinfo [title author])
(defrecord Dvdinfo  [title star region])
(defrecord Gameinfo [title])

(defmulti visit (fn[visitor info] [(class visitor) (class info)]))

(defmethod visit [Shortblurb Bookinfo] [visitor info]
	(str "SB-Book: " (:title info)))
(defmethod visit [Longblurb  Bookinfo] [visitor info]
	(str "LB-Book: " (:title info)
		", Author: " (:author info)))
	
(defmethod visit [Shortblurb Gameinfo] [visitor info]
	(str "SB-Game: " (:title info)))
(defmethod visit [Longblurb  Gameinfo] [visitor info]
	(str "LB-Game: " (:title info)))

(defmethod visit [Shortblurb Dvdinfo] [visitor info]
	(str "SB-DVD: " (:title info)))
(defmethod visit [Longblurb  Dvdinfo] [visitor info]
	(str "LB-DVD: " (:title info) 
		", starring " (:star info) 
		", encoding region: " (:region info)))

(defn accept [info blurb]
	(visit blurb info))
